Bluetooth: Mesh: Add Multi Advertising Sets Support

LingaoM opened the pull request:
This PR add multi advertising sets for bluetooth mesh adv bearer.
At the present, the local mesh network occupy single advertising.
And GATT Services Advertising use a single advertising, if support.

One or more advertising sets maybe used by relay node.

Signed-off-by: Lingao Meng <menglingao@xiaomi.com>

LingaoM committed:
Bluetooth: Mesh: Add Multi Advertising Sets Support

This PR add multi advertising sets for bluetooth mesh adv bearer.
At the present, the local mesh network occupy single advertising.
And GATT Services Advertising use a single advertising, if support.

One or more advertising sets maybe used by relay node.

Signed-off-by: Lingao Meng <menglingao@xiaomi.com>

LingaoM commented:
@trond-snekvik FYI

LingaoM committed:
Bluetooth: Mesh: Add build multi extened advertising test

Add new testcase to tests/bluetooth/mesh that builds mesh with
multi advertising sets.

Signed-off-by: Lingao Meng <menglingao@xiaomi.com>

tejlmand commented on subsys/bluetooth/mesh/Kconfig:
```diff
@@ -299,6 +299,15 @@ config BT_MESH_ADV_EXT
+	help
+	  Use multi extended advertising commands for operating the advertiser.
+	  Multi extended advertising is faster than legacy or single extended
+	  advertising, but isn't supported by all controllers.
```
> but isn't supported by all controllers.

How does that piece of information help the reader ?
I assume the above `depends on` settings ensures that this setting can only be enabled if the currently selected controller supports `BT_MESH_ADV_EXT_MULTI` or do we need some additional Kconfig settings so that proper dependency chain can be ensured.

tejlmand commented on subsys/bluetooth/mesh/Kconfig:
```diff
@@ -513,6 +522,16 @@ config BT_MESH_RELAY_RETRANSMIT_INTERVAL
+	int "Maximum of simultaneous relay message support"
+	depends on BT_MESH_ADV_EXT_MULTI
+	depends on BT_EXT_ADV_MAX_ADV_SET > 1
+	default 1
```
a range will default to the lowest value in the range if not specified.

So you can remove this.
```suggestion
```

tejlmand commented on subsys/bluetooth/mesh/Kconfig:
```diff
@@ -513,6 +522,16 @@ config BT_MESH_RELAY_RETRANSMIT_INTERVAL
+config BT_MESH_RELAY_SIMULTANEITY_COUNT
+	int "Maximum of simultaneous relay message support"
+	depends on BT_MESH_ADV_EXT_MULTI
+	depends on BT_EXT_ADV_MAX_ADV_SET > 1
```
any particular reason this has a dependency to `BT_EXT_ADV_MAX_ADV_SET > 1`.

In the range below, if `BT_EXT_ADV_MAX_ADV_SET == 1` it will be:
```
	range 1 1
```
which is perfectly valid range,  where only `1` is a valid value.

tejlmand commented on tests/bluetooth/mesh/multi_ext_adv.conf:
```diff
@@ -0,0 +1,64 @@
+CONFIG_BT_MESH_DEBUG=y
+CONFIG_BT_MESH_DEBUG_PROV=y
+CONFIG_BT_MESH_DEBUG_PROXY=y
+#CONFIG_BT_MESH_DEBUG_BEACON=y
```
Why is this commented out ?

tejlmand commented on tests/bluetooth/mesh/multi_ext_adv.conf:
```diff
@@ -0,0 +1,64 @@
+CONFIG_BT_MESH_DEBUG_PROXY=y
+#CONFIG_BT_MESH_DEBUG_BEACON=y
+CONFIG_BT_MESH_DEBUG_NET=y
+#CONFIG_BT_MESH_DEBUG_TRANS=y
+#CONFIG_BT_MESH_DEBUG_LOW_POWER=y
+#CONFIG_BT_MESH_DEBUG_FRIEND=y
+#CONFIG_BT_MESH_DEBUG_MODEL=y
```
why is this commented out ?

tejlmand commented on tests/bluetooth/mesh/multi_ext_adv.conf:
```diff
@@ -0,0 +1,64 @@
+
+CONFIG_BT_MESH_IV_UPDATE_TEST=y
+
+#CONFIG_UART_CONSOLE=n
+#CONFIG_BT_DEBUG_MONITOR_UART=y
```
why is this commented out.

tejlmand reviewed (commented):
some minor build system related observations

LingaoM commented on subsys/bluetooth/mesh/adv_legacy.c:
```diff
@@ -123,6 +123,55 @@ static inline void adv_send(struct net_buf *buf)
+	};
+
+	for (int i = 0; i < ARRAY_SIZE(events); i++) {
+		events[i].state = K_POLL_STATE_NOT_READY;
```
That true.

LingaoM replied:
poll will check current fifo at begin: https://github.com/zephyrproject-rtos/zephyr/blob/main/kernel/poll.c#L72

LingaoM commented on subsys/bluetooth/mesh/adv.h:
```diff
@@ -58,6 +59,9 @@ struct net_buf *bt_mesh_adv_create_from_pool(struct net_buf_pool *pool,
 		      void *cb_data);
 
+void bt_mesh_relay_send(struct net_buf *buf, const struct bt_mesh_send_cb *cb,
+		      void *cb_data);
```
`tag` is good idea.

LingaoM committed:
Bluetooth: Mesh: Move unused public api

Move bt_mesh_adv_create_from_pool to source code

Signed-off-by: Lingao Meng <menglingao@xiaomi.com>

LingaoM committed:
Bluetooth: Mesh: Introduce adv tag field when create buf

Introduce tag field when create adv buffer.

Signed-off-by: Lingao Meng <menglingao@xiaomi.com>

Andrewpini commented on subsys/bluetooth/mesh/adv_multi.c:
```diff
@@ -0,0 +1,393 @@
+		return;
+	}
+
+	for (; adv_ins <= BT_MESH_RELAY_ADV_INS_END; adv_ins++) {
```
```suggestion
	for (; adv_ins <= (struct ext_adv*)BT_MESH_RELAY_ADV_INS_END; adv_ins++) {
```

<warning: comparison of distinct pointer types lacks a cast>

LingaoM replied:
Is it necessary? Or did you encounter this warning?

Andrewpini replied:
Got this warning while compiling the changes

LingaoM replied:
OK. thanks, i will rebase.

Andrewpini commented:
I have done some measurements on a desktop setup, and it seems like this PR can resolve one of the challenges for large scale networks. From previous work I have seen that relay nodes quickly become bottlenecks in larger networks where messages are continuously generated at a moderate or high rate, an issue that multiple advertising sets is likely to remedy.

I have one proposal to the PR. In the current implementation there is one dedicated advertising set that handles locally generated advertising for the device. Is it possible to create a solution where the local advertising can utilize multiple advertising sets as well?  I believe this will enhance the peak transmission rate of the device, using the same principles as for the relays. 

Possible solutions to this could be to let local and relay messages share a common pool of advertising sets, alternatively configuring the number of local/relay adv sets at compile time.

LingaoM commented:
> I have done some measurements on a desktop setup, and it seems like this PR can resolve one of the challenges for large scale networks. From previous work I have seen that relay nodes quickly become bottlenecks in larger networks where messages are continuously generated at a moderate or high rate, an issue that multiple advertising sets is likely to remedy.
> 
> 
> 
> I have one proposal to the PR. In the current implementation there is one dedicated advertising set that handles locally generated advertising for the device. Is it possible to create a solution where the local advertising can utilize multiple advertising sets as well?  I believe this will enhance the peak transmission rate of the device, using the same principles as for the relays. 
> 
> 
> 
> Possible solutions to this could be to let local and relay messages share a common pool of advertising sets, alternatively configuring the number of local/relay adv sets at compile time.

Hi

We cannot use multiple advertising sets for local mesh messages, because every mesh message will contain orders sequence number. It's must be send by sequence. Otherwise will be replay by peer node.

@trond-snekvik Any comments for this?

Andrewpini commented:
> Hi
> 
> We cannot use multiple advertising sets for local mesh messages, because every mesh message will contain orders sequence number. It's must be send by sequence. Otherwise will be replay by peer node.
> 
> @trond-snekvik Any comments for this?

Ok, I see the Issue. So multiple adv sets does not guarantee in-order transmission of outgoing messages, resulting in loss of messages due to replay protection. I have talked to @trond-snekvik and this issue seems legit 👍

LingaoM commented on subsys/bluetooth/mesh/adv_multi.c:
```diff
@@ -0,0 +1,393 @@
+#else /* !CONFIG_BT_MESH_RELAY */
+#define RELAY_ADV_COUNT			0
+#define BT_MESH_RELAY_ADV_INS_START	(struct ext_adv *)NULL
+#define BT_MESH_RELAY_ADV_INS_END	(struct ext_adv *)NULL
```
@Andrewpini CC

trond-snekvik commented on subsys/bluetooth/mesh/adv.c:
```diff
@@ -84,16 +88,66 @@ struct net_buf *bt_mesh_adv_create_from_pool(struct net_buf_pool *pool,
+
+	for (int i = 0; i < ARRAY_SIZE(events); i++) {
+		events[i].state = K_POLL_STATE_NOT_READY;
+	}
```
Thanks for the explanation on this. Would it make sense to put the poll events on stack instead, and initialize them with `k_poll_event_init`? Should save 80 byte of static ram in adv_ext mode, since that won't max out the stack depth

LingaoM replied:
Looks Make sense to me

trond-snekvik commented on subsys/bluetooth/mesh/adv_ext.c:
```diff
@@ -155,7 +155,7 @@ static void send_pending_adv(struct k_work *work)
-	while ((buf = net_buf_get(&bt_mesh_adv_queue, K_NO_WAIT))) {
+	while ((buf = bt_mesh_adv_buf_get(K_NO_WAIT))) {
 		/* busy == 0 means this was canceled */
 		if (!BT_MESH_ADV(buf)->busy) {
```
This `busy` logic can be moved inside `bt_mesh_adv_buf_get` now

LingaoM replied:
Compared with the modification of existing logic and keep assist with legacy with ext, it has little significance.

trond-snekvik commented on subsys/bluetooth/mesh/adv_legacy.c:
```diff
@@ -178,9 +177,18 @@ void bt_mesh_adv_update(void)
+
+#if defined(CONFIG_BT_MESH_RELAY)
+	k_fifo_cancel_wait(&bt_mesh_relay_queue);
+#endif /* CONFIG_BT_MESH_RELAY */
```
Should put this inside adv.c as well.

LingaoM replied:
Where?

trond-snekvik replied:
I mean, since the polling of the queues are implemented in adv.c, it makes sense to put the cancellation of the polling inside that file as well, as a new `bt_mesh_adv_buf_get_cancel()` function or something

LingaoM replied:
Maybe, but adv_multi.c really need cancel fifo, since use k_no_wait.

trond-snekvik commented on subsys/bluetooth/mesh/CMakeLists.txt:
```diff
@@ -23,7 +23,13 @@ zephyr_library_sources_ifdef(CONFIG_BT_MESH
 zephyr_library_sources_ifdef(CONFIG_BT_MESH_ADV_LEGACY adv_legacy.c)
 
-zephyr_library_sources_ifdef(CONFIG_BT_MESH_ADV_EXT adv_ext.c)
+if(CONFIG_BT_MESH_ADV_EXT)
+if (CONFIG_BT_MESH_RELAY_ADV_SETS)
+zephyr_library_sources(adv_multi.c)
+else()
+zephyr_library_sources(adv_ext.c)
+endif()
+endif()
```
I don't think my comment was very clear. Can adv_multi.c replace adv_ext.c completely? There appears to be quite a bit of duplicate logic in there

LingaoM replied:
I am worried that the replacement will have some impact. For example, now adv_multi.c is different from adv_ext.c in processing proxy advertising. The former has an exclusive advertising set.

LingaoM replied:
To complete the replacement, two additional tasks need to be completed:

- Supporting proxy advertising, occupying a separate broadcast set or a shared advertising set.
- Support relay to occupy a separate broadcast set or shared advertising set.

LingaoM replied:
But complete replacement is meaningful because it avoids the existence of duplicate code and is conducive to further development.

Andrewpini reviewed (approved):
LGTM 👍 I think it was the right call to merge adv_ext and adv_multi into a single module. It will make it easier to manage the adv layer in the future.

trond-snekvik commented on subsys/bluetooth/mesh/adv_multi.c:
```diff
@@ -0,0 +1,452 @@
+	if (adv_param.interval_min != BT_MESH_ADV_SCAN_UNIT(adv_int)) {
+		adv_param.interval_min = BT_MESH_ADV_SCAN_UNIT(adv_int);
+		adv_param.interval_max = adv_param.interval_min;
+	}
```
IMO, it'd be safer to keep `adv_param` on stack now that it's used in so many parallel advertisers. Perhaps the previously configured interval could be stored inside the `ext_adv` struct, so we can keep the old behavior of not updating advertising parameters when the interval is unchanged?

trond-snekvik commented on subsys/bluetooth/mesh/adv_multi.c:
```diff
@@ -0,0 +1,452 @@
+			return; /* Wait for advertising to finish */
+		}
+	}
+}
```
It should be possible to get rid of this duplication between `send_local_adv` and `send_relay_adv`. 

 Could for instance change the `tag` to a bitfield enum (that is, ADV_LOCAL is `BIT(0)` and ADV_RELAY is `BIT(1)`), then make the `bt_mesh_adv_buf_get` function take a `tag` parameter, which fetches from one or both of the queues based on the indicated tags.

This would also remove the last remaining reference to `bt_mesh_relay_queue`and `bt_mesh_local_queue` outside the adv.c file, and they would no longer have to be `extern`ed.

trond-snekvik commented on subsys/bluetooth/mesh/Kconfig:
```diff
@@ -301,6 +301,26 @@ config BT_MESH_ADV_EXT
+	  Maximum of simultaneous relay message support. Requires controller support
+	  multiple advertising sets.
+
+config BT_MESH_GATT_ADV
```
How does GATT provisioning work in adv_multi.c without this? If it's possible to set this to `n` while `GATT_PROXY` is `y`, we have to be able to send proxy beacons on the main advertiser

trond-snekvik commented on subsys/bluetooth/mesh/adv_multi.c:
```diff
@@ -0,0 +1,452 @@
+		delta = k_uptime_delta(&adv_ins->timestamp);
+		k_work_reschedule(&adv_ins->work, K_MSEC(ADV_INT_FAST_MS - delta));
+
+	}
```
This duplication should also be removed

trond-snekvik reviewed (changes requested):
IMO, with the combined functionality, adv_multi.c should be called adv_ext.c, as that's what the corresponding kconfig symbol is named.

LingaoM commented on subsys/bluetooth/mesh/adv_ext.c:
```diff
@@ -51,62 +43,133 @@ enum {
+static void send_pending_adv(struct k_work *work);
+static struct ext_adv adv_main = {
+	.work = Z_WORK_DELAYABLE_INITIALIZER(send_pending_adv),
+	.handler = bt_mesh_adv_buf_get,
```
Here, there is a change. The first advertising set (call `adv_main`) has the ability to send relay messages. I think this is also reasonable. When there is no local message to send, it should not choose to ignore it.

PavelVPV commented on subsys/bluetooth/mesh/adv_ext.c:
```diff
@@ -51,62 +43,133 @@ enum {
 	struct bt_le_ext_adv *instance;
 	struct net_buf *buf;
 	uint64_t timestamp;
 	struct k_work_delayable work;
```
I feel that only one `k_work_delayable` would be enough for all advertiser. Each advertiser could only keep the next scheduling time and the work could be scheduled at the earliest possible scheduling time. Although the code complexity may increase a bit.

LingaoM replied:
It will be complicated, but I think it can be put in the future PR.

PavelVPV commented on subsys/bluetooth/mesh/Kconfig:
```diff
@@ -301,6 +301,26 @@ config BT_MESH_ADV_EXT
+	  multiple advertising sets.
+
+config BT_MESH_GATT_ADV
+	bool "Use a separate extended advertising set for GATT Server Advertising"
```
May be rename it to `BT_MESH_ADV_EXT_GATT_SEPARATE` or so? The current name doesn't describe the purpose of this option.

LingaoM commented on subsys/bluetooth/mesh/adv_ext.c:
```diff
@@ -172,122 +257,208 @@ static void send_pending_adv(struct k_work *work)
-	if (!IS_ENABLED(CONFIG_BT_MESH_GATT_SERVER)) {
-		return;
+	if (IS_ENABLED(CONFIG_BT_MESH_GATT_SERVER) &&
+	    adv == BT_MESH_GATT_ADV_INS) {
```
No, also include relay advertising set.

LingaoM committed:
Bluetooth: Mesh: add static bt_mesh_adv_create_from_pool

Since this function not use by other place, so add static.

Signed-off-by: Lingao Meng <menglingao@xiaomi.com>

LingaoM committed:
Bluetooth: Mesh: Rename gatt adv func

Since this module is already changed, current really
confuses when trying to understand the code.

Signed-off-by: Lingao Meng <menglingao@xiaomi.com>

LingaoM commented on subsys/bluetooth/mesh/adv_ext.c:
```diff
@@ -172,122 +257,208 @@ static void send_pending_adv(struct k_work *work)
+
+	do {
+		if (schedule_send(adv++)) {
+			return;
```
Every time call `bt_mesh_adv_send` with tag = relay, will call `bt_mesh_adv_buf_relay_ready`, so if there two relay message send, will call twice.

I don't understand that there will be. When there are two or more in this queue, only one advertising set works.

LingaoM committed:
Bluetooth: Mesh: Add macro for bt_mesh_relay_queue

If multi relay advertising feature not used, we can
compiled out for save some memory.

Signed-off-by: Lingao Meng <menglingao@xiaomi.com>

LingaoM committed:
Bluetooth: Mesh: Move common gatt adv to adv.c

Since gatt adv send have common logic between adv_ext.c
and adv_legacy.c, so move to adv.c

Signed-off-by: Lingao Meng <menglingao@xiaomi.com>

trond-snekvik commented on subsys/bluetooth/mesh/adv.c:
```diff
@@ -31,14 +31,17 @@
 
-K_FIFO_DEFINE(bt_mesh_adv_queue);
+static K_FIFO_DEFINE(bt_mesh_adv_queue);
+static K_FIFO_DEFINE(bt_mesh_relay_queue);
```
If the rest of the code doesn't reference this in anything that's compiled in, it doesn't need the `#if` wrapper, the linker will just silently remove it.

trond-snekvik commented on subsys/bluetooth/mesh/adv_ext.c:
```diff
@@ -51,62 +42,133 @@ enum {
+#else /* !CONFIG_BT_MESH_ADV_EXT_GATT_SEPARATE */
+#define BT_MESH_ADV_COUNT			(1 + RELAY_ADV_COUNT)
+#define BT_MESH_ADV_EXT_GATT_SEPARATE_INS	(struct ext_adv)NULL
+#endif /* CONFIG_BT_MESH_ADV_EXT_GATT_SEPARATE */
```
Please make this a static inline function instead.
https://www.kernel.org/doc/html/v4.10/process/coding-style.html#macros-enums-and-rtl
> Generally, inline functions are preferable to macros resembling functions.

trond-snekvik commented on subsys/bluetooth/mesh/adv_ext.c:
```diff
@@ -51,62 +42,133 @@ enum {
+#else /* !CONFIG_BT_MESH_RELAY */
+#define RELAY_ADV_COUNT				0
+#define BT_MESH_RELAY_ADV_INS			(struct ext_adv)NULL
+#endif /* CONFIG_BT_MESH_RELAY */
```
This should also be a static inline function

trond-snekvik commented on subsys/bluetooth/mesh/adv_ext.c:
```diff
@@ -51,62 +42,133 @@ enum {
+};
+
+#if defined(CONFIG_BT_MESH_RELAY)
+#define RELAY_ADV_COUNT			CONFIG_BT_MESH_RELAY_ADV_SETS
```
Since this is just a plain alias now, it'd be easier to follow if we just used `CONFIG_BT_MESH_RELAY_ADV_SETS` throughout

trond-snekvik commented on subsys/bluetooth/mesh/adv_ext.c:
```diff
@@ -51,62 +42,133 @@ enum {
 };
 
-static struct {
+typedef struct net_buf* (*adv_buf_handler_t)(k_timeout_t timeout);
```
```suggestion
typedef struct net_buf* (*adv_buf_alloc_t)(k_timeout_t timeout);
```
with a corresponding name change in the `ext_adv` struct below.

This could also be handled by replacing the two separate tag specific adv_get and buf_ready functions with a single one that takes LOCAL/RELAY tags as a bitfield. Could even add a third GATT_PROXY tag, and all the if/else business surrounding the allocation of advertising instances depending on configuration would go away, and the buffer selection would just be determined by what tags an advertiser instance has. 

There's currently a lot of `#if/#else` code here that doesn't have to be that separated if we accessed allocation buffers based on parameters, rather than functions.

This could also be achieved by adding "role" flags to the existing `ext_adv` `flags` field. The logic for assign the flags could be done in the init function to keep it simple. I would argue that you wouldn't even need three different sets of advertiser instances, just make one pool, and assign the "role" tags to the appropriate ones, depending on the configuration.

LingaoM replied:
>  I would argue that you wouldn't even need three different sets of advertiser instances, just make one pool, and assign the "role" tags to the appropriate ones, depending on the configuration.

https://github.com/zephyrproject-rtos/zephyr/pull/40031#discussion_r750300086

Add section for this.

trond-snekvik commented on subsys/bluetooth/mesh/adv_ext.c:
```diff
@@ -51,62 +42,133 @@ enum {
+
+#if defined(CONFIG_BT_MESH_RELAY)
+#define RELAY_ADV_COUNT			CONFIG_BT_MESH_RELAY_ADV_SETS
+#if RELAY_ADV_COUNT
```
If you remove this `#if`-wrapper, you don't have to protect all access to `adv_relay` in preprocessor wrappers. An array of length 0 is permitted in gcc, and will allocate 0 bytes in the final image

trond-snekvik commented on subsys/bluetooth/mesh/adv_ext.c:
```diff
@@ -51,62 +42,133 @@ enum {
+#if defined(CONFIG_BT_MESH_DEBUG_USE_ID_ADDR)
+		.options = BT_LE_ADV_OPT_USE_IDENTITY,
+#endif
+	}
```
Since this is identical in every structure, perhaps it's easier to just do this in init?

trond-snekvik reviewed (changes requested):
Sorry for keeping this PR going, but this is a major change to the very foundation of the mesh stack, and I don't want to let any changes through that would cause problems down the line. 

I think there are some things that could be made a lot simpler here by reducing the amount of preprocessor directives and using the tags to map queues to instances, rather than separate functions.

LingaoM committed:
Bluetooth: Mesh: Introduce tag field

Introduce tag field.

Signed-off-by: Lingao Meng <menglingao@xiaomi.com>

LingaoM committed:
Bluetooth: Mesh: Add common section for ext_adv

Use section for omit macro config.

Signed-off-by: Lingao Meng <menglingao@xiaomi.com>

LingaoM committed:
Bluetooth: Mesh: Replace macro as inline

Use inline function ins of macro.

Signed-off-by: Lingao Meng <menglingao@xiaomi.com>

trond-snekvik reviewed (changes requested):
LGTM, but please squash this into some more appropriate commits

trond-snekvik commented:
@LingaoM are each of the commits in a usable state? We'd like to be able to bisect the main branch in a way that we can always build and run the stack on each commit.

LingaoM commented:
> @LingaoM are each of the commits in a usable state? We'd like to be able to bisect the main branch in a way that we can always build and run the stack on each commit. 

Yes，every commit can always build and run.

LingaoM commented:
@jhedberg CC

jhedberg commented on subsys/bluetooth/mesh/adv.h:
```diff
@@ -25,40 +25,49 @@ enum bt_mesh_adv_type {
 		  started:1,
-		  busy:1;
+		  busy:1,
+		  tag:1;
```
This cannot work. You’re reserving a single bit for a value that needs at least three bits.

LingaoM replied:
``` C
enum bt_mesh_adv_tag {
	BT_MESH_LOCAL_ADV,
	BT_MESH_RELAY_ADV,

	/* Only used with bitfield */
	BT_MESH_PROXY_ADV,
};

```

Actually, we don't use `BT_MESH_PROXY_ADV` when call `bt_mesh_adv_buf_send`();

LingaoM replied:
In order to avoid misunderstanding, I specially added a note `	/* Only used with bitfield */
 `.

jhedberg replied:
What got me confused is that you're defining two sets of values for the different types of packets. One set is made up of mutually orthogonal values (using the `BIT()` macro) that can be logically or-ed together, and another is an enum. I thought you were assigning the `BIT()` values to the one bit wide bitfield.

jhedberg replied:
What would help, I think, is to have a comment for this bit field that only the enum values are assigned to it, and not the `BT_MESH_ADV_TAG_*` values. The comment you're referring to in the enum definition didn't really help in anyway to clarify this detail.

LingaoM replied:
If `BIT() `definitions are used, it will occupy two extra bits. I just hope to leave more space for the future` uint8_t      type:2,`

Maybe I need to ask for @trond-snekvik  advice.

jhedberg replied:
Alternatively, since there is space for it, extend the width of the field to three bits and completely remove the confusing duplicate enumeration, i.e. only use the `BIT()` values.

LingaoM replied:
Done.

jhedberg reviewed (changes requested):
Unless I’m missing something, there’s a pretty fundamental bug with how much memory is allocated for storing the tag value.

LingaoM commented on subsys/bluetooth/mesh/adv.c:
```diff
@@ -60,10 +64,11 @@ static struct bt_mesh_adv *adv_alloc(int id)
+static struct net_buf *bt_mesh_adv_create_from_pool(struct net_buf_pool *pool,
+						    bt_mesh_adv_alloc_t get_id,
+						    enum bt_mesh_adv_type type,
+						    enum bt_mesh_adv_tag tag,
```
@jhedberg `enum bt_mesh_adv_tag tag` only used `BT_MESH_LOCAL_ADV` and `BT_MESH_RELAY_ADV`, so 1 bit enough.

jhedberg commented on subsys/bluetooth/mesh/adv.c:
```diff
@@ -84,16 +89,103 @@ struct net_buf *bt_mesh_adv_create_from_pool(struct net_buf_pool *pool,
+
+struct net_buf *bt_mesh_adv_buf_get_by_tag(uint8_t tag, k_timeout_t timeout)
+{
+	if (!(tag ^ BT_MESH_ADV_TAG_RELAY)) {
```
I really had to wrap my head around this too long, and I'm still not sure if I got the intention right. The more common way of checking if a specific bit (and nothing else) is set is `if ((val & MY_BIT) == MY_BIT)`. That said, I assume the purpose you're exclusively checking for that bit is that if anything else is set then that has higher priority. If that's the case, why don't you just order the if-statements so that the relay case becomes an `else if` statement?

LingaoM replied:
Great.

jhedberg commented on subsys/bluetooth/mesh/adv.h:
```diff
@@ -25,40 +25,44 @@ enum bt_mesh_adv_type {
+	BT_MESH_LOCAL_ADV = BIT(0),
+	BT_MESH_RELAY_ADV = BIT(1),
+
+	/* Only used with bitfield */
```
Eliminating the secondary enumeration makes this look clearer now, thanks. However, I realised that the usage of the "bitfield" term is what caused a lot of the confusion to begin with. A bitfield (at least the way I use the term) is a variable with an explicitly stated width in bits, e.g. `uint8_t foo:2;`. And that's exactly what you're *not* using the proxy value for. I think the more appropriate term would be "bitmask" or referring to the `bt_mesh_adv.tag` struct member. However, I think you could just remove the code comment completely at this point.

LingaoM replied:
Sorry, I don’t quite understand your expression, what do I need to do?

trond-snekvik replied:
@LingaoM 
> remove the code comment completely at this point.

jhedberg commented on subsys/bluetooth/mesh/adv_ext.c:
```diff
@@ -172,122 +220,175 @@ static void send_pending_adv(struct k_work *work)
+		if (schedule_send(adv++)) {
+			return;
+		}
+	} while (CONFIG_BT_MESH_RELAY_ADV_SETS > ++i);
```
Could you just convert this to the more usual for-loop?
```
for (int i = 0; i < CONFIG_BT_MESH_RELAY_ADV_SETS; i++) {
        if (schedule_send(&adv[i])) {
                return;
        }
}
```

I've used `&adv[i]` there instead of `adv++` since iterating a pointer you got from some function return looks dangerous. It's good to make it clear that you're treating this pointer as an array whose valid indices follow the same bounds as the `i` variable.
